File: swarm/ts/workstreams/herdr.ts
import type { ExtensionAPI } from "@earendil-works/pi-coding-agent";
import { errorMessage } from "../agents/errors.ts";

export const HERDR_WORKSTREAM_OPEN_TIMEOUT_MS = 5000;

export interface HerdrWorkstreamEnv {
	HERDR_ENV?: string;
	HERDR_SOCKET_PATH?: string;
	HERDR_PANE_ID?: string;
	HERDR_WORKSPACE_ID?: string;
	BASECAMP_AGENT_DEPTH?: string;
}

export interface HerdrWorkstreamWorkspaceInput {
	protectedRoot?: string;
	repo?: {
		root?: string;
	};
	launchCwd?: string;
	hasUI?: boolean;
}

export interface HerdrWorkstreamWorktreeInput {
	path: string;
	label: string;
}

export type HerdrWorkstreamSkipReason =
	| "missing-herdr-env"
	| "missing-herdr-socket-path"
	| "missing-herdr-pane-id"
	| "subagent"
	| "headless"
	| "missing-cwd";

export interface HerdrWorkstreamOpenArgsOpened {
	args: string[];
}

export interface HerdrWorkstreamOpenArgsSkipped {
	args: null;
	status: "skipped";
	reason: HerdrWorkstreamSkipReason;
	message: string;
}

export type HerdrWorkstreamOpenArgsResult = HerdrWorkstreamOpenArgsOpened | HerdrWorkstreamOpenArgsSkipped;

export type HerdrWorkstreamOpenResult =
	| {
			status: "opened";
			message: string;
			args: string[];
			stdout?: string;
			stderr?: string;
	  }
	| {
			status: "skipped";
			reason: HerdrWorkstreamSkipReason;
			message: string;
	  }
	| {
			status: "failed";
			message: string;
			error?: string;
			exitCode?: number;
			stdout?: string;
			stderr?: string;
			args?: string[];
	  };

export interface HerdrWorkstreamOpenEligibilityInput {
	env: HerdrWorkstreamEnv;
	hasUI?: boolean;
}

function skipped(reason: HerdrWorkstreamSkipReason, message: string): HerdrWorkstreamOpenArgsSkipped {
	return { args: null, status: "skipped", reason, message };
}

function agentDepth(env: HerdrWorkstreamEnv): number {
	const raw = env.BASECAMP_AGENT_DEPTH;
	if (raw === undefined || raw.trim() === "") return 0;
	const parsed = Number(raw);
	return Number.isFinite(parsed) ? parsed : 1;
}

export function shouldOpenWorkstreamInHerdr(
	input: HerdrWorkstreamOpenEligibilityInput,
): HerdrWorkstreamOpenArgsSkipped | null {
	if (input.env.HERDR_ENV !== "1")
		return skipped("missing-herdr-env", "Herdr workstream open skipped: not running in Herdr.");
	if (!input.env.HERDR_SOCKET_PATH) {
		return skipped("missing-herdr-socket-path", "Herdr workstream open skipped: missing Herdr socket path.");
	}
	if (!input.env.HERDR_PANE_ID) {
		return skipped("missing-herdr-pane-id", "Herdr workstream open skipped: missing Herdr pane id.");
	}
	if (agentDepth(input.env) !== 0) {
		return skipped("subagent", "Herdr workstream open skipped: only primary sessions can open workstreams in Herdr.");
	}
	if (input.hasUI === false) {
		return skipped("headless", "Herdr workstream open skipped: session has no UI.");
	}
	return null;
}

function workspaceCwd(workspace: HerdrWorkstreamWorkspaceInput): string | null {
	return workspace.protectedRoot ?? workspace.repo?.root ?? workspace.launchCwd ?? null;
}

export function buildHerdrWorkstreamOpenArgs(
	workspace: HerdrWorkstreamWorkspaceInput,
	worktree: HerdrWorkstreamWorktreeInput,
	env: HerdrWorkstreamEnv,
): HerdrWorkstreamOpenArgsResult {
	const skip = shouldOpenWorkstreamInHerdr({ env, hasUI: workspace.hasUI });
	if (skip) return skip;

	const args = ["worktree", "open"];
	if (env.HERDR_WORKSPACE_ID) {
		args.push("--workspace", env.HERDR_WORKSPACE_ID);
	} else {
		const cwd = workspaceCwd(workspace);
		if (!cwd) return skipped("missing-cwd", "Herdr workstream open skipped: missing workspace cwd.");
		args.push("--cwd", cwd);
	}
	args.push("--path", worktree.path, "--label", worktree.label, "--no-focus", "--json");
	return { args };
}

export async function openWorkstreamInHerdr(
	pi: Pick<ExtensionAPI, "exec">,
	workspace: HerdrWorkstreamWorkspaceInput,
	worktree: HerdrWorkstreamWorktreeInput,
	env: HerdrWorkstreamEnv = process.env,
): Promise<HerdrWorkstreamOpenResult> {
	const built = buildHerdrWorkstreamOpenArgs(workspace, worktree, env);
	if (built.args === null) {
		return { status: "skipped", reason: built.reason, message: built.message };
	}

	try {
		const result = await pi.exec("herdr", built.args, { timeout: HERDR_WORKSTREAM_OPEN_TIMEOUT_MS });
		if (result.code !== 0) {
			return {
				status: "failed",
				message: `Herdr workstream open failed with exit code ${result.code}.`,
				exitCode: result.code,
				stdout: result.stdout,
				stderr: result.stderr,
				args: built.args,
			};
		}
		return {
			status: "opened",
			message: "Herdr workstream opened.",
			args: built.args,
			stdout: result.stdout,
			stderr: result.stderr,
		};
	} catch (err) {
		return {
			status: "failed",
			message: "Herdr workstream open failed.",
			error: errorMessage(err),
			args: built.args,
		};
	}
}

File: src/basecamp/claude/herdr.py
"""Open a workstream worktree in a Herdr pane (Python parallel of the Pi adapter).

Best-effort by design: a failed or absent ``herdr`` never breaks staging — the
record and worktree are already valid, so a pane failure is reported and the user
opens the pane by hand. Ports the Pi ``herdr.ts`` eligibility predicate verbatim
and shells out to the ``herdr`` CLI (a plain PATH binary).
"""

from __future__ import annotations

import subprocess
from collections.abc import Mapping
from dataclasses import dataclass, field

_HERDR_OPEN_TIMEOUT_S = 5.0


@dataclass(frozen=True)
class HerdrResult:
    """Outcome of a Herdr pane open: ``opened`` | ``skipped`` | ``failed``."""

    status: str
    message: str
    reason: str | None = None
    args: tuple[str, ...] = field(default_factory=tuple)


def _agent_depth(env: Mapping[str, str]) -> int:
    raw = env.get("BASECAMP_AGENT_DEPTH")
    if raw is None or raw.strip() == "":
        return 0
    try:
        return int(float(raw))
    except ValueError:
        return 1


def herdr_skip_reason(env: Mapping[str, str], *, has_ui: bool = True) -> str | None:
    """Return a skip reason if a Herdr pane cannot/should not be opened, else ``None``.

    Ports ``shouldOpenWorkstreamInHerdr``: needs ``HERDR_ENV=1`` +
    ``HERDR_SOCKET_PATH`` + ``HERDR_PANE_ID``, a primary (depth-0) session, and a UI.
    """

    if env.get("HERDR_ENV") != "1":
        return "missing-herdr-env"
    if not env.get("HERDR_SOCKET_PATH"):
        return "missing-herdr-socket-path"
    if not env.get("HERDR_PANE_ID"):
        return "missing-herdr-pane-id"
    if _agent_depth(env) != 0:
        return "subagent"
    if has_ui is False:
        return "headless"
    return None


def build_open_args(
    *,
    worktree_path: str,
    label: str,
    workspace_cwd: str | None,
    env: Mapping[str, str],
) -> list[str] | None:
    """Build the ``herdr worktree open`` argv, or ``None`` if a required cwd is missing.

    Prefers ``--workspace <HERDR_WORKSPACE_ID>``; falls back to ``--cwd`` when the
    workspace id is absent. Caller has already checked :func:`herdr_skip_reason`.
    """

    args = ["worktree", "open"]
    workspace_id = env.get("HERDR_WORKSPACE_ID")
    if workspace_id:
        args += ["--workspace", workspace_id]
    else:
        if not workspace_cwd:
            return None
        args += ["--cwd", workspace_cwd]
    args += ["--path", worktree_path, "--label", label, "--no-focus", "--json"]
    return args


def open_pane(
    *,
    worktree_path: str,
    label: str,
    workspace_cwd: str | None,
    env: Mapping[str, str],
    has_ui: bool = True,
) -> HerdrResult:
    """Best-effort open ``worktree_path`` in a Herdr pane. Never raises."""

    reason = herdr_skip_reason(env, has_ui=has_ui)
    if reason is not None:
        return HerdrResult(status="skipped", message=f"Herdr pane open skipped: {reason}.", reason=reason)

    args = build_open_args(worktree_path=worktree_path, label=label, workspace_cwd=workspace_cwd, env=env)
    if args is None:
        return HerdrResult(status="skipped", message="Herdr pane open skipped: missing cwd.", reason="missing-cwd")

    try:
        result = subprocess.run(
            ["herdr", *args],
            capture_output=True,
            text=True,
            timeout=_HERDR_OPEN_TIMEOUT_S,
            check=False,
        )
    except FileNotFoundError:
        return HerdrResult(status="skipped", message="Herdr pane open skipped: herdr not on PATH.", reason="no-herdr")
    except (OSError, subprocess.SubprocessError) as exc:
        return HerdrResult(status="failed", message=f"Herdr pane open failed: {exc}.", args=tuple(args))
    if result.returncode != 0:
        return HerdrResult(
            status="failed",
            message=f"Herdr pane open failed with exit code {result.returncode}.",
            args=tuple(args),
        )
    return HerdrResult(status="opened", message="Herdr pane opened.", args=tuple(args))
